formats/cbor/commonMain/src/kotlinx/serialization/cbor/ByteStringWrapper.kt

```diff
@@ -0,0 +1,86 @@
+package kotlinx.serialization.cbor
+
+import kotlinx.serialization.*
+import kotlinx.serialization.builtins.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.encoding.*
+
+/**
+ * Use this class if you'll need to serialize a complex type as a byte string before encoding it,
+ * i.e. as it is the case with the protected header in COSE structures.
+ *
+ * An example for a COSE header data class would be:
+ *
+ * ```
+ * @Serializable
+ * data class CoseHeader(
+ *     @CborLabel(1)
+ *     @SerialName("alg")
+ *     val alg: Int? = null
+ * )
+ *
+ * @Serializable
+ * data class CoseSigned(
+ *     @ByteString
+ *     @CborLabel(1)
+ *     @SerialName("protectedHeader")
+ *     val protectedHeader: ByteStringWrapper<CoseHeader>,
+ * )
+ * ```
+ *
+ * Serializing this `CoseHeader` object would result in `a10143a10126`, in diagnostic notation:
+ *
+ * ```
+ * A1           # map(1)
+ *    01        # unsigned(1)
+ *    43        # bytes(3)
+ *       A10126 # "\xA1\u0001&"
+ * ```
+ *
+ * so the `protectedHeader` got serialized first and then encoded as a `@ByteString`.
+ *
+ * Note that `equals()` and `hashCode()` only use `value`, not `serialized`.
+ */
+@Serializable(with = ByteStringWrapperSerializer::class)
+public class ByteStringWrapper<T>(
+    public val value: T,
+    public val serialized: ByteArray = byteArrayOf()
```

Comment:
I will just add my thoughts, feel free to ignore, but I believe that `ByteStringWrapper` should not be included in `kotlinx-serialization-cbor`.
Why?
1. because it has rather specific use case with possibility to create inconsistent data in it
2. it's rather easy to implement it outside of `kotlinx-serialization-cbor`
3. as stated in https://github.com/Kotlin/kotlinx.serialization/pull/2412#issuecomment-2190011858, serializer currently uses default `Cbor` instance, which could be not what user expects and to reuse the same `Cbor` configuration we will need to expose `CborDecoder`/`CborEncoder` public interfaces (like in `Json`)

So I have only one question: why it should be provided out of the box in `kotlinx-serialization-cbor`?

Reply:
It does need fixing (point 3).
Would you mind clarifying? If it is straight-forward to implement outside `kotlinx-serialization-cbor`, but fixing the issue requires exposing what is currently internal (`CborDecoder`/`CborEncoder`), how can it be implemented outside?

I get your concern about inconsistent data, but I don't agree it is overly specific. If you want to work with CBOR Web Tokens you will hit a situation rather sooner than later where you need this specific functionality (again, COSE is an absolutely horrible specification and the world would be a better place without it. Yet here we are…)

Reply:
On a more general note. Is there a way to achieve the desired behaviour without risking inconsistent data?

Reply:
@whyoleg  Thank you for bringing this to my attention. Initially I thought that ByteStringWrapper is required for Cbor implementation (because it is handled in special way in `beginStructure`), but now I see it is not used there or anywhere else in the CborWriter or CborReader. So, thoretically it would be possible to implement it separately. The question is whether it is a popular use-case enough to provide it out-of-the-box — so people won't have to copy-paste the implementation.
Other concerns are still valid though. The way I see it, there are two options:

### 1. ByteStringWrapper won't be included

In this case, to properly reuse Cbor settings, `CborReader/Writer` would have to implement new interfaces, like this one:
```kotlin
interface CborEncoder {
    val cbor: Cbor
}
```

It is not that hard to add them though.

### 2. ByteStringWrapper is included

In this case, additional work should also be done:
1. Instead of creating CborEncoder/Decoder, it is possible to cast `Encoder` to internal `CborWriter` to extract settings from there — to fix the bug where settings are not taken into account
2. It should be added to guide (`formats.md`) — currently it is not mentioned there.
3. `val serialized: ByteArray` needs rework. I mentioned that it is possible to create invalid data by calling constructor like `ByteStringWrapper("foo", byteArrayOf(0x0))`. My proposal is the following: make `serialized` nullable and make constructor that accepts it internal, leaving to public constructor that accepts only `value`. I.e.:
```kotlin
 public class ByteStringWrapper<T> internal constructor(
   public val value: T,  
   public val serialized: ByteArray?
 ) {
    public constuctor(value: T): this(value, null)
 }
```

By doing so, we can be sure that `serialized` was produced only by `ByteStringWrapperSerializer` and will be null in user-constructed objects.

Alternatively, we can require a Cbor instance to create ByteStringWrapper:

```kotlin
 public class ByteStringWrapper<T> internal constructor(
     public val value: T,  
     public val serialized: ByteArray
 ) {
    
 }

public inline fun <T> ByteStringWrapper(cbor: Cbor, value: T) = ByteStringWrapper(value, cbor.encodeToByteArray(value))
// or even
fun <T> Cbor.toByteStringWrapper(value: T) = ByteStringWrapper(value, encodeToByteArray(value))
```

Since this approach will perform serialization on every constructor call, I'd recommend to avoid it unless you really need to have non-null `serialized` value for manually created instances.
Of course you can also combinate these two approaches in case you want to have them both.
But the thing is, you should be able to pass an arbitrary ByteArray to create ByteStringWrapper.

Reply:
> It does need fixing (point 3). Would you mind clarifying? If it is straight-forward to implement outside `kotlinx-serialization-cbor`, but fixing the issue requires exposing what is currently internal (`CborDecoder`/`CborEncoder`), how can it be implemented outside?

Yeah, you’re right. I was intended to write that it is better to provide minimal `CborDecoder`/`CborEncoder` with access to `Cbor` instance instead of `ByteStringWrapper` class itself.

> I get your concern about inconsistent data, but I don't agree it is overly specific. If you want to work with CBOR Web Tokens you will hit a situation rather sooner than later where you need this specific functionality (again, COSE is an absolutely horrible specification and the world would be a better place without it. Yet here we are…)

Yeah, that's what I'm curious — is it needed somewhere outside COSE/CWT?
I mean, all other changes are following CBOR specifications, but `ByteStringWrapper`, looks a bit out of place. I would imaging, that there are use-cases for `Json` to have `StringWrapper` (JWT?), but we don't have such class in `kotlinx-serialization-json`.

So I would propose to exclude it and provide minimal `CborDecoder`/`CborEncoder` as stated above and create an issue, if needed, to collect other use-cases.

Reply:
As an additional note: `kotlinx-io` now has `ByteString` class, so both `@ByteString` and `ByteStringWrapper` are now somehow looks confusing (as they control how `ByteArray` is serialized).
May be we need to rename `@ByteString` to `@CborByteString` to avoid confusion? (same for `ByteStringWrapper` -> `CborByteStringWrapper`)

Should I create an issue for this?

Reply:
You have convinced me by making a good case with the JOSE/JWT analogy and from this point of view, I agree that it is not required and should go to a COSE-specific library. As we are maintaining such a library anyhow, excluding it here is a non-issue for us, as long as the proposed public `CborEncoder` interface is introduced.

I will say, however, with CBOR and COSE being what they are that I am certain anyone implementing COSE will implement something like this from scratch. Of course, this is fine, because this library deals with serialization in general and not COSE specifically. It is very possible that this issue will pop up at some point in the future. For the sake of completeness, I'd like to mention that we had to come up with something similar for validating CWTs in our multiplatform codebase and I would not want that part of kotlinix.serialization. 
I also really like the implementation suggestions and I'll shamelessly take inspiration from them for our codebase